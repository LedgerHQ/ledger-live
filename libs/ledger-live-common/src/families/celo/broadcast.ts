import { AccountBridge } from "@ledgerhq/types-live";
import { patchOperationWithHash } from "../../operation";
import { Transaction } from "./types";
import { celoKit } from "./api/sdk";

export const broadcast: AccountBridge<Transaction>["broadcast"] = async ({
  signedOperation: { operation, signature },
}) => {
  const kit = celoKit();
  const { transactionHash } = await kit.web3.eth.sendSignedTransaction(signature);
  return patchOperationWithHash(operation, transactionHash);
};

export default broadcast;
