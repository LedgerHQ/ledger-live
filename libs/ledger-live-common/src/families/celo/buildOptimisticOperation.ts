import BigNumber from "bignumber.js";
import { OperationType } from "@ledgerhq/types-live";
import { CeloAccount, CeloOperation, CeloOperationMode, Transaction } from "./types";
import { encodeOperationId } from "../../operation";

const MODE_TO_TYPE: { [key in CeloOperationMode | "default"]: OperationType } = {
  send: "OUT",
  lock: "LOCK",
  unlock: "UNLOCK",
  withdraw: "WITHDRAW",
  vote: "VOTE",
  revoke: "REVOKE",
  activate: "ACTIVATE",
  register: "REGISTER",
  default: "FEES",
};

export const buildOptimisticOperation = (
  account: CeloAccount,
  transaction: Transaction,
  fee: BigNumber,
): CeloOperation => {
  const type = MODE_TO_TYPE[transaction.mode] ?? MODE_TO_TYPE.default;

  const value =
    type === "OUT" || type === "LOCK"
      ? new BigNumber(transaction.amount).plus(fee)
      : new BigNumber(transaction.amount);

  const operation: CeloOperation = {
    id: encodeOperationId(account.id, "", type),
    hash: "",
    type,
    value,
    fee,
    blockHash: null,
    blockHeight: null,
    senders: [account.freshAddress],
    recipients: [transaction.recipient].filter(Boolean),
    accountId: account.id,
    date: new Date(),
    extra: {
      celoOperationValue: new BigNumber(transaction.amount),
      ...(["ACTIVATE", "VOTE", "REVOKE"].includes(type)
        ? {
            celoSourceValidator: transaction.recipient,
          }
        : {}),
    },
  };

  return operation;
};
