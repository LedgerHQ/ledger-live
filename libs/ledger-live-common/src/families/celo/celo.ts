import Eth from "@ledgerhq/hw-app-eth";
import {
  tokenInfoByAddressAndChainId,
  legacyTokenInfoByAddressAndChainId,
} from "@celo/wallet-ledger/lib/tokens";

import SemVer from "semver";

export default class Celo extends Eth {
  // celo-spender-app below version 1.2.3 used a different private key to validate erc20 token info.
  // this legacy version of the app also only supported celo type 0 transactions.
  // if you are reading this after celo moved to op based L2 those celo type 0 transactions will no longer work
  // so you can safely remove all the legacy paths.
  async verifyTokenInfo(to: string, chainId: number): Promise<void> {
    const isModern = await this.isAppModern();
    const tokenInfo = isModern
      ? tokenInfoByAddressAndChainId(to!, chainId!)
      : legacyTokenInfoByAddressAndChainId(to!, chainId!);

    if (tokenInfo) {
      // celo-spender-app below version 1.2.3 expected unprefixed hex strings only
      const dataString = isModern
        ? `0x${tokenInfo.data.toString("hex")}`
        : tokenInfo.data.toString("hex");

      await this.provideERC20TokenInformation(dataString);
    }
  }

  private async isAppModern(): Promise<boolean> {
    const { version } = await this.getAppConfiguration();

    return SemVer.statisfy(version, ">= 1.2.3");
  }
}
