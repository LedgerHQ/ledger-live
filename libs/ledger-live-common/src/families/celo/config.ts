export * from "@ledgerhq/coin-celo/config";
