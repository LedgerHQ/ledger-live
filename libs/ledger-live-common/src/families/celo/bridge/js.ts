import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { sync, scanAccounts } from "../js-synchronisation";
import { getPreloadStrategy, preload, hydrate } from "../preload";
import signOperation from "../js-signOperation";
import getTransactionStatus from "../js-getTransactionStatus";
import broadcast from "../js-broadcast";
import estimateMaxSpendable from "../js-estimateMaxSpendable";
import prepareTransaction from "../js-prepareTransaction";
import createTransaction from "../js-createTransaction";

const updateTransaction = (t, patch) => ({ ...t, ...patch });

const receive = makeAccountBridgeReceive();

const currencyBridge: CurrencyBridge = {
  getPreloadStrategy,
  preload,
  hydrate,
  scanAccounts,
};
const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};
export default {
  currencyBridge,
  accountBridge,
};
