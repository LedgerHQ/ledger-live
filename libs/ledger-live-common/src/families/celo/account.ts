import { BigNumber } from "bignumber.js";
import { CeloOperationExtra, CeloOperationExtraRaw } from "./types";

export function fromOperationExtraRaw(extraRaw: CeloOperationExtraRaw): CeloOperationExtra {
  const extra: CeloOperationExtra = {
    celoOperationValue: new BigNumber(extraRaw.celoOperationValue),
  };

  if (extraRaw.celoSourceValidator) {
    extra.celoSourceValidator = extraRaw.celoSourceValidator;
  }

  return extra;
}

export function toOperationExtraRaw(extra: CeloOperationExtra): CeloOperationExtraRaw {
  const extraRaw: CeloOperationExtraRaw = {
    celoOperationValue: extra.celoOperationValue.toString(),
  };

  if (extra.celoSourceValidator) {
    extraRaw.celoSourceValidator = extra.celoSourceValidator;
  }

  return extraRaw;
}

export default {
  fromOperationExtraRaw,
  toOperationExtraRaw,
};
