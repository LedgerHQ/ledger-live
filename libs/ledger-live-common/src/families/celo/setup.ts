// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-celo/bridge";
import makeCliTools from "@ledgerhq/coin-celo/cli-transaction";
import CeloResolver from "@ledgerhq/coin-celo/hw-getAddress";
import Transport from "@ledgerhq/hw-transport";
import type { Bridge } from "@ledgerhq/types-live";
import Celo from "@ledgerhq/hw-app-celo";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";
import { Transaction, CeloAccount } from "./types";
import { CeloSigner } from "@ledgerhq/coin-celo/signer";

const createSigner: CreateSigner<CeloSigner> = (transport: Transport) => new Celo(transport);

const bridge: Bridge<Transaction, CeloAccount> = createBridges(executeWithSigner(createSigner));

const resolver: Resolver = createResolver(createSigner, CeloResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
