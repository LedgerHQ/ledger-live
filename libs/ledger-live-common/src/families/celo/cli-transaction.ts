import type {
  Account,
  AccountLike,
  AccountLikeArray,
} from "@ledgerhq/types-live";

import { from } from "rxjs";
import { map } from "rxjs/operators";
import { getValidatorGroups } from "../celo/api";
import invariant from "invariant";
import flatMap from "lodash/flatMap";
import type { Transaction } from "../celo/types";

const options = [
  {
    name: "mode",
    type: String,
    desc: "mode of transaction: send, lock, unlock, withdraw, vote, revoke, activate, register",
  },
  {
    name: "transactionIndex",
    type: String,
    desc: "transaction index of a pending withdraw in case of withdraw mode",
  },
];

function inferAccounts(account: Account): AccountLikeArray {
  invariant(account.currency.family === "celo", "celo family");

  const accounts: Account[] = [account];
  return accounts;
}

function inferTransactions(
  transactions: Array<{
    account: AccountLike;
    transaction: Transaction;
    mainAccount: Account;
  }>,
  opts: Record<string, any>
): Transaction[] {
  const mode = opts.mode || "send";
  invariant(
    [
      "send",
      "lock",
      "unlock",
      "withdraw",
      "vote",
      "revoke",
      "activate",
      "register",
    ].includes(mode),
    `Unexpected mode: ${mode}`
  );

  return flatMap(transactions, ({ transaction }) => {
    invariant(transaction.family === "celo", "celo family");

    return {
      ...transaction,
      family: "celo",
      mode,
      index: opts.transactionIndex || null,
    } as Transaction;
  });
}

const celoValidatorGroups = {
  args: [],
  job: () =>
    from(getValidatorGroups()).pipe(
      map((validatorGroup) => JSON.stringify(validatorGroup))
    ),
};

export default {
  options,
  inferAccounts,
  inferTransactions,
  commands: {
    celoValidatorGroups,
  },
};
