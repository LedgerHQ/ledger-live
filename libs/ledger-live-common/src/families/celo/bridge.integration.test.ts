import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import celo from "./datasets/celo.scanAccounts.1";
import type { Transaction } from "./types";
import type { DatasetTest } from "@ledgerhq/types-live";

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    celo,
  },
};

testBridge(dataset);
