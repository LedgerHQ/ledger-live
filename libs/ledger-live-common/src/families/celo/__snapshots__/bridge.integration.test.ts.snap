// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`celo currency bridge scanAccounts Celo 1 1`] = `
Array [
  Object {
    "balance": "200081719000000000",
    "celoResources": Object {
      "electionAddress": "0x8D6677192144292870907E3Fa8A5527fE55A7ff6",
      "lockedBalance": "200000000000000000",
      "lockedGoldAddress": "0x6cC083Aed9e3ebe302A6336dBC7c921C9f03349E",
      "maxNumGroupsVotedFor": "10",
      "nonvotingLockedBalance": "100000000000000000",
      "pendingWithdrawals": Array [],
      "registrationStatus": true,
      "votes": Array [
        Object {
          "activatable": true,
          "amount": "50000000000000000",
          "index": 0,
          "revokable": true,
          "type": "pending",
          "validatorGroup": "0x0861a61Bf679A30680510EcC238ee43B82C5e843",
        },
        Object {
          "activatable": true,
          "amount": "50000000000000000",
          "index": 0,
          "revokable": true,
          "type": "pending",
          "validatorGroup": "0x35aE10F412503aBcf9275133613E8df7f56E72Be",
        },
      ],
    },
    "currencyId": "celo",
    "derivationMode": "",
    "freshAddress": "0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB",
    "freshAddressPath": "44'/52752'/0'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB",
        "derivationPath": "44'/52752'/0'/0/0",
      },
    ],
    "id": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:",
    "index": 0,
    "name": "Celo 1",
    "nfts": undefined,
    "operationsCount": 7,
    "pendingOperations": Array [],
    "seedIdentifier": "0453390dcc1e6f1be0fb34f837b278ed1b4c84097c7493c13a0d915c735af1d8aa445d2738a95e8f50bc22abd2e17cb868d4db22a623d99d861740eb93373d50a3",
    "spendableBalance": "81719000000000",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 18,
    "used": true,
  },
  Object {
    "balance": "0",
    "celoResources": Object {
      "electionAddress": "0x8D6677192144292870907E3Fa8A5527fE55A7ff6",
      "lockedBalance": "0",
      "lockedGoldAddress": "0x6cC083Aed9e3ebe302A6336dBC7c921C9f03349E",
      "maxNumGroupsVotedFor": "10",
      "nonvotingLockedBalance": "0",
      "pendingWithdrawals": Array [],
      "registrationStatus": false,
      "votes": Array [],
    },
    "currencyId": "celo",
    "derivationMode": "",
    "freshAddress": "0x2cC7E5913bADa8FA2895bDeF7F4C2E36C2368Abb",
    "freshAddressPath": "44'/52752'/1'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "0x2cC7E5913bADa8FA2895bDeF7F4C2E36C2368Abb",
        "derivationPath": "44'/52752'/1'/0/0",
      },
    ],
    "id": "js:2:celo:0x2cC7E5913bADa8FA2895bDeF7F4C2E36C2368Abb:",
    "index": 1,
    "name": "Celo 2",
    "nfts": undefined,
    "operationsCount": 0,
    "pendingOperations": Array [],
    "seedIdentifier": "0453390dcc1e6f1be0fb34f837b278ed1b4c84097c7493c13a0d915c735af1d8aa445d2738a95e8f50bc22abd2e17cb868d4db22a623d99d861740eb93373d50a3",
    "spendableBalance": "0",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 18,
    "used": false,
  },
]
`;

exports[`celo currency bridge scanAccounts Celo 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:",
      "blockHash": null,
      "blockHeight": 14356282,
      "contract": undefined,
      "extra": Object {
        "celoOperationValue": "150000000000000000",
      },
      "fee": "33936000000000",
      "hasFailed": false,
      "hash": "0x13b90cebe150eadbebd525200f2ff7ae8e838defbee31a9812a802356705b0f2",
      "id": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:-0x13b90cebe150eadbebd525200f2ff7ae8e838defbee31a9812a802356705b0f2-LOCK",
      "operator": undefined,
      "recipients": Array [
        "0x6cc083aed9e3ebe302a6336dbc7c921c9f03349e",
      ],
      "senders": Array [
        "0xcfd48e0faf9f19377509ce68a6a6f4d9c85ff8ab",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "LOCK",
      "value": "33936000000000",
    },
    Object {
      "accountId": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:",
      "blockHash": null,
      "blockHeight": 14356260,
      "contract": undefined,
      "extra": Object {
        "celoOperationValue": "0",
      },
      "fee": "23628500000000",
      "hasFailed": false,
      "hash": "0x1bedc04795ee8c4b8739e661498ad7721f5f0f8dc957a7e2009314d447be20a8",
      "id": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:-0x1bedc04795ee8c4b8739e661498ad7721f5f0f8dc957a7e2009314d447be20a8-REGISTER",
      "operator": undefined,
      "recipients": Array [
        "0x7d21685c17607338b313a7174bab6620bad0aab7",
      ],
      "senders": Array [
        "0xcfd48e0faf9f19377509ce68a6a6f4d9c85ff8ab",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "REGISTER",
      "value": "23628500000000",
    },
    Object {
      "accountId": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:",
      "blockHash": null,
      "blockHeight": 14356275,
      "contract": undefined,
      "extra": Object {
        "celoOperationValue": "50000000000000000",
        "celoSourceValidator": "0x0861a61bf679a30680510ecc238ee43b82c5e843",
      },
      "fee": "104239000000000",
      "hasFailed": false,
      "hash": "0x3f8609e7d154aac9d4165d0587ecbf058dcd45a9b86d313558c0066069d8ef0d",
      "id": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:-0x3f8609e7d154aac9d4165d0587ecbf058dcd45a9b86d313558c0066069d8ef0d-VOTE",
      "operator": undefined,
      "recipients": Array [
        "0x0861a61bf679a30680510ecc238ee43b82c5e843",
      ],
      "senders": Array [
        "0xcfd48e0faf9f19377509ce68a6a6f4d9c85ff8ab",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "VOTE",
      "value": "104239000000000",
    },
    Object {
      "accountId": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:",
      "blockHash": null,
      "blockHeight": 14345376,
      "contract": undefined,
      "extra": Object {
        "celoOperationValue": "399893159500000000",
      },
      "fee": "18281000000000",
      "hasFailed": false,
      "hash": "0x6aa0cc933bfcf828bad4bfc16f9474b287aa754ec37f19563724ea60c1c2e2da",
      "id": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:-0x6aa0cc933bfcf828bad4bfc16f9474b287aa754ec37f19563724ea60c1c2e2da-IN",
      "operator": undefined,
      "recipients": Array [
        "0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB",
      ],
      "senders": Array [
        "0x330be4C2B77f0f08f037c1D04788D00344765556",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "399893159500000000",
    },
    Object {
      "accountId": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:",
      "blockHash": null,
      "blockHeight": 14391317,
      "contract": undefined,
      "extra": Object {
        "celoOperationValue": "199497368500000000",
      },
      "fee": "18281000000000",
      "hasFailed": false,
      "hash": "0x9d8263a088cf91dd3376b2c4a033a51401414441cf747fde724901b4480b563f",
      "id": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:-0x9d8263a088cf91dd3376b2c4a033a51401414441cf747fde724901b4480b563f-OUT",
      "operator": undefined,
      "recipients": Array [
        "0xDE680aC69f10F8B7e8B547f6da866369D059ac81",
      ],
      "senders": Array [
        "0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "OUT",
      "value": "199497368500000000",
    },
    Object {
      "accountId": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:",
      "blockHash": null,
      "blockHeight": 14356307,
      "contract": undefined,
      "extra": Object {
        "celoOperationValue": "50000000000000000",
        "celoSourceValidator": "0x35ae10f412503abcf9275133613e8df7f56e72be",
      },
      "fee": "100051500000000",
      "hasFailed": false,
      "hash": "0xb1d95aa993e5175effba1eec6aa7e5a557c5e1d6f8c3d4414fbe6e206b6f8841",
      "id": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:-0xb1d95aa993e5175effba1eec6aa7e5a557c5e1d6f8c3d4414fbe6e206b6f8841-VOTE",
      "operator": undefined,
      "recipients": Array [
        "0x35ae10f412503abcf9275133613e8df7f56e72be",
      ],
      "senders": Array [
        "0xcfd48e0faf9f19377509ce68a6a6f4d9c85ff8ab",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "VOTE",
      "value": "100051500000000",
    },
    Object {
      "accountId": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:",
      "blockHash": null,
      "blockHeight": 14356267,
      "contract": undefined,
      "extra": Object {
        "celoOperationValue": "50000000000000000",
      },
      "fee": "33936000000000",
      "hasFailed": false,
      "hash": "0xd00974d8cb71e48ac9a3085579c9b589a4d67cec348aa61b9b48cc7b3bcd54bf",
      "id": "js:2:celo:0xcfD48e0FAf9f19377509cE68a6A6F4D9C85ff8AB:-0xd00974d8cb71e48ac9a3085579c9b589a4d67cec348aa61b9b48cc7b3bcd54bf-LOCK",
      "operator": undefined,
      "recipients": Array [
        "0x6cc083aed9e3ebe302a6336dbc7c921c9f03349e",
      ],
      "senders": Array [
        "0xcfd48e0faf9f19377509ce68a6a6f4d9c85ff8ab",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "LOCK",
      "value": "33936000000000",
    },
  ],
  Array [],
]
`;
